Add basic request tracing facility

tbg opened the pull request:
For discussion:
added the `tracer` package, providing facilities to trace requests through the cluster.

reworked API based on the suggestions from the first proposal: a Trace is
autonomous, and submits itself on `Finalize()`. For tracing requests, have
- `trace.Event("x happened")` and
- `fn := trace.Epoch("stuff begins now"); fn() // stuff ends now`, with its<br />
  most convenient use case `defer trace.Epoch("x")()`.

Epochs and Events generate the same entries, but the former with nonzero
duration. Each entry has a depth, so nested epochs can be displayed nicely:

single-request increment, node side. Note that the total duration of the call
(609.166us) is right at the top; the nested epochs below divvy it up.

```
ID                 Origin              Ts                 Dur          Desc                    File
1435427552[...]    macts.local:8001    13:52:32.200487    609.166µs    node                    server/node.go:564
1435427552[...]    macts.local:8001    13:52:32.200515    576.619µs    ·executing Increment    storage/store.go:1271
1435427552[...]    macts.local:8001    13:52:32.200554    538.349µs    ··range                 storage/range.go:517
1435427552[...]    macts.local:8001    13:52:32.200561    0            ···write path           storage/range.go:526
1435427552[...]    macts.local:8001    13:52:32.200576    515.994µs    ···raft                 storage/range.go:703
1435427552[...]    macts.local:8001    13:52:32.200583    48.923µs     ····proposing           storage/range.go:765
1435427552[...]    macts.local:8001    13:52:32.200971    107.385µs    ····executing           storage/range.go:799
```

`EndTransaction` on the client gateway side:

```
I0627 13:52:32.785017   53305 monitor.go:211] received trace:
ID          Origin              Ts                 Dur            Desc                             File
2f2ecd58    macts.local:8001    13:52:32.743443    11.079554ms    coordinated sender               kv/txn_coord_sender.go:324
2f2ecd58    macts.local:8001    13:52:32.743451    1.57359ms      ·sending EndTransaction          kv/dist_sender.go:638
2f2ecd58    macts.local:8001    13:52:32.743455    5.607µs        ··meta descriptor lookup         kv/dist_sender.go:675
2f2ecd58    macts.local:8001    13:52:32.743483    0              ··sending to macts.local:8001    rpc/send.go:178
```

Transactional `Get` which hits an intent, attempts to resolve it
asynchronously, backs off and gets an error not retryable at the store level.

```
ID          Origin              Ts                 Dur             Desc                   File
3de447e3    macts.local:8001    13:52:32.888115    139.456068ms    node                   server/node.go:564
3de447e3    macts.local:8001    13:52:32.888125    126.494154ms    ·executing Get         storage/store.go:1271
3de447e3    macts.local:8001    13:52:32.888128    110.367783ms    ··range                storage/range.go:517
3de447e3    macts.local:8001    13:52:32.888131    110.364629ms    ···read path           storage/range.go:523
3de447e3    macts.local:8001    13:52:32.998509    16.098773ms     ··intent resolution    storage/store.go:1343
3de447e3    macts.local:8001    13:52:33.014457    150.888µs       ···range               storage/range.go:517
3de447e3    macts.local:8001    13:52:33.014460    0               ····write path         storage/range.go:526
3de447e3    macts.local:8001    13:52:33.014489    0               ····raft [async]       storage/range.go:698
3de447e3    macts.local:8001    13:52:33.014625    165.802µs       ·backoff               storage/store.go:1264
3de447e3    macts.local:8001    13:52:33.014796    12.765902ms     ·executing Get         storage/store.go:1271
3de447e3    macts.local:8001    13:52:33.014802    12.759787ms     ··range                storage/range.go:517
3de447e3    macts.local:8001    13:52:33.014805    12.755809ms     ···read path           storage/range.go:523
3de447e3    macts.local:8001    13:52:33.027566    0               ·not successful        storage/store.go:1266
```

TODO:
- convert to protos
- metadata handling? (ex: peer at request boundary)
- figure out persisting and viewing Traces (as discussed in PR)

tschottdorf commented:
some related reading:

* Cassandra offers [probabilistic tracing](http://docs.datastax.com/en/cassandra/2.1/cassandra/tools/toolsSetTraceProbability.html) and manual [per-request tracing](http://docs.datastax.com/en/cql/3.1/cql/cql_reference/tracing_r.html); the result is a list of events with source node and elapsed time (I suspect that their events have a beginning and an end to measure their time elapsed, whereas my proposal measures the time from one event to the next - probably worth to improving that. The traces are written to an extra table, one trace incurring "at least 10 writes". They expire after 24 hours.
* MongoDB can either profile everything (not recommended obviously), slow queries or nothing. Profiling doesn't seem to do well with shards. "Slow" queries are those which take longer than a specified threshold, defaulting to `100ms`. Traces are written to a local ring buffer.

For ease of access it sounds desirable to store traces in Cockroach (replicated, to keep UIs symmetrical), but of course our writes are expensive compared to Cassandra, and a ring buffer on MVCC isn't great. If we limit the trace keyspace to inline writes (assuming we won't need transactions there) and have each node batch up its traces (dropping when it goes over a threshold - nobody should be tracing huge amounts, but it can happen) it seems doable.

spencerkimball commented:
As far as writing to Cockroach, I think it's a good idea to
probabilistically trace so we can keep our writes to a tolerable level. We
should also support manual tracing. Were you thinking to have a protobuf
format and support merges based on trace ID? Or just multiple records for
distributed traces with key being something like trace ID + node ID?

On Tue, Jun 23, 2015 at 12:49 PM, Tobias Schottdorf <
notifications@github.com> wrote:

> some related reading:
>
>    - Cassandra offers probabilistic tracing
>    <http://docs.datastax.com/en/cassandra/2.1/cassandra/tools/toolsSetTraceProbability.html>
>    and manual per-request tracing
>    <http://docs.datastax.com/en/cql/3.1/cql/cql_reference/tracing_r.html>;
>    the result is a list of events with source node and elapsed time (I suspect
>    that their events have a beginning and an end to measure their time
>    elapsed, whereas my proposal measures the time from one event to the next -
>    probably worth to improving that. The traces are written to an extra table,
>    one trace incurring "at least 10 writes". They expire after 24 hours.
>    - MongoDB can either profile everything (not recommended obviously),
>    slow queries or nothing. Profiling doesn't seem to do well with shards.
>    "Slow" queries are those which take longer than a specified threshold,
>    defaulting to 100ms. Traces are written to a local ring buffer.
>
> For ease of access it sounds desirable to store traces in Cockroach
> (replicated, to keep UIs symmetrical), but of course our writes are
> expensive compared to Cassandra, and a ring buffer on MVCC isn't great. If
> we limit the trace keyspace to inline writes (assuming we won't need
> transactions there) and have each node batch up its traces (dropping when
> it goes over a threshold - nobody should be tracing huge amounts, but it
> can happen) it seems doable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1446#issuecomment-114568851>
> .
>

mrtracy commented on util/feed.go:
```diff
@@ -91,6 +91,9 @@ type Feed struct {
 // all Subscribers to the feed. Events published to a closed feed, or to a feed
 // with no Subscribers, will be ignored.
 func (f *Feed) Publish(event interface{}) {
+	if f == nil {
```
Why get rid of this?  In my mind, the feed has valid nil behavior; the behavior of a nil feed is the same as a non-nil feed with zero subscribers.

tschottdorf replied:
Not sure I understand, using a `nil` feed would previously panic. Now it behaves closer to a closed feed.
This edit isn't part of some grand plan yet, but it would be convenient if a nil feed behaved like a dummy feed that doesn't do anything (but you can still subscribe to it). Otherwise various places need to do some dancing around this exceptional case. StoreEventFeed checks this all the time, for instance.
How do you feel about that?

mrtracy replied:
I misread this diff, because I was 100% positive I had added this behavior before and it highlighted because it was being taken out.  Reading it now, it's clear that its being added; my brain is clearly broken.  

The nil behavior that you have added is the correct thing to do, disregard my comment.

tschottdorf replied:
ok. You have a bunch of those checks in `{server/status,storage}/feed.go`, those should be safe to remove when this lands.

mrtracy replied:
Okay, that's where I had added them.  It's coming back to me.

mrtracy commented on util/tracer.go:
```diff
@@ -0,0 +1,276 @@
+type Trace struct {
+	// ID is the unique identifier for this group of requests.
+	ID string
+	// Content is a trace, containing call sites and timings in the order in
```
"Content contains call sites..."

mrtracy commented on util/tracer.go:
```diff
@@ -0,0 +1,276 @@
+}
+
+// A Tracer is used to follow requests across the system (or across systems).
+// Requests must implement the TraceID interface and can be traced by invoking
```
Traceable interface

mrtracy commented on util/tracer.go:
```diff
@@ -0,0 +1,276 @@
+		t.traces[id] = append(t.traces[id], TraceItem{
+			Pos:       pos,
+			Func:      fun,
+			Timestamp: t.now(),
```
timestamp outside of loop?

bdarnell commented:
At google all request traces were written into in-memory ring buffers and then a subset were persisted to disk (and the subset was chosen deterministically based on the trace ID so the same request would be persisted or not across all servers that handled it). There were multiple ring buffers for different timing thresholds (all requests, > 10ms, > 100ms, > 500ms, > 1s) so that outliers would be kept in memory for longer (I think the slow requests were also logged but separately from the random sample). This was extremely useful; I probably used the local in-memory version more often than the distributed version because the sample rate for the distributed storage was so low.

For a UI, it's apparently possible to feed external events into the chrome dev tools; square did this for tracing network activity in ios apps: https://github.com/square/PonyDebugger. I've also extracted a similar UI from the app engine SDK: https://github.com/bdarnell/tornado_tracing

mrtracy commented:
I'm wondering if this shouldn't be factored a bit differently with regards to feeds.

Right now a pointer to the tracer is being passed around, which accumulates data internally until a finalize is called, which publishes a feed.  The way I would think of feeds, the Begin(), Trace() and Finalize() functions would just publish "TraceItems" to the feed, and a higher-level component would assemble them into traces.  In fact, I don't think you would even need Begin() or Finalize() events in that case; you could just trace everything with an ID, and the higher-level system could just periodically flush everything it knows about a particular ID to disk.

bdarnell commented:
I like the Begin/Finalize pairing as an interface to make sure we get start and end times for everything we want to time (and Finalize could be a cue to explicitly trigger a flush). 

I'm not sure it makes sense to combine tracing and feeds. Some events certainly should trigger both (e.g. the raft events this PR adds to range.go), but I think it would be cleaner to call both tracer.Trace and feed.Publish in that case instead of having the higher-level system recognize that *this* TraceItem should also be a timeseries data point.

bdarnell commented on kv/dist_sender.go:
```diff
@@ -184,6 +186,10 @@ func NewDistSender(ctx *DistSenderContext, gossip *gossip.Gossip) *DistSender {
 	}
+	ds.tracer = ctx.Tracer
+	if ds.tracer == nil {
+		ds.tracer = util.NewTracer(nil, 2)
```
It seems strange to create a Tracer here (same in store.go). Since the tracer values tolerate nil methods why not just leave it as nil unless a value was passed in?

tschottdorf replied:
This is going away, I added this before I patched up the Feed. The intention is that this can be nil as you suggest.

bdarnell commented on kv/dist_sender.go:
```diff
@@ -650,6 +656,7 @@ func (ds *DistSender) Send(_ context.Context, call proto.Call) {
 
 	curReply := finalReply
 	for {
+		ds.tracer.Trace(args.Header(), "ds_send")
```
Is it legal to call Trace without Begin? I think I'd like to see all Trace calls happen within a Begin/Finalize pair (this would give them a place to exist in a timeline-based UI). Maybe Begin() should return an object with Trace and Finalize methods instead of making all 3 methods on Tracer (this way the args.Header() parameter wouldn't need to be repeated).

tschottdorf replied:
It's currently legal, but rare. A `Trace` without a `Begin` will immediately finalize (so you get a single-item trace chunk). So as long as a request passed at least one Begin/Finalize block, it'll be in that block, otherwise it's its own Trace (example: commands popping out of Raft). That seems like a good way of handling things to me.

bdarnell commented on kv/txn_coord_sender.go:
```diff
@@ -309,6 +311,10 @@ func (tc *TxnCoordSender) Send(_ context.Context, call proto.Call) {
 	tc.maybeBeginTxn(header)
 
+	header.CmdID = header.GetOrCreateCmdID(tc.clock.PhysicalNow())
+	tc.tracer.Begin(header, "coordsender_begin")
```
The suffixes "_begin" and "_done" should be added by Begin and Finalize to ensure they're consistent everywhere.

tschottdorf replied:
I'm going to do some more work on the interface. Begin and Finalize are called when a request enters/leaves the system, respectively. They should attach the origin/location of where the request may go next, and I'll have to figure out whether we should continue allowing multiple calls to `Begin()` (they are counted against calls to `Finalize()`; this is necessary/useful since `DistSender` and `Store` share a tracer to combine as much of the trace as possible). Likely Finalize will lose its message. BTW, there's no special relationship between the Begin and Finalize calls in the final trace (i.e. no diff between them is stored directly, diffs are always between consecutive traced events), just in case that impression came up.

bdarnell commented on storage/store.go:
```diff
@@ -315,6 +315,9 @@ type StoreContext struct {
 	EventFeed *util.Feed
+
+	// Tracer is a request tracer used by this store for request tracing.
+	Tracer *util.Tracer
```
Could we add the Tracer to the context.Context instead of threading another parameter throughout the codebase?

tschottdorf replied:
I'll play around with the API. Probably a Trace object which already knows its TraceID as you suggest above is a good candidate for putting into the context.

tnachen commented:
@bdarnell And what sort of information do the traces gather from Google besides just timing?

bdarnell commented:
@tnachen: Freeform logging. `tracer.Trace()` was often a substitute for `VLOG(1)`. Trace logs were like VLOGs in that they were cheap and not seen by default, but they were exposed whenever there was a slow request instead of being controlled by a global flag.

mrtracy commented:
I think it would actually be good factoring for the higher-level module to know which trace events are being recorded as time series.  The feed system was created to support separation of concerns; the lower-level components publish a simple feed of ordered events with some context, and it's left to the higher level status components to determine how to interpret and persist that data.

tschottdorf commented:
@spencerkimball: Merge op would work but I'm not going to go there until there's a clear benefit for simplicity. `timestamp+node` sounds like the natural one to start with.

MBerhault commented:
A few general comments:
- we'll probably want a trace GC at some point. For now though, a hook into the stopper so that we can catch non-finalized traces in tests.
- the tracer seem too synchronized. How about changing it to channels to add new traces, and only modify the map on the channel consumer?

tschottdorf commented:
@MBerhault good idea with the hook. Can you elaborate on "too synchronized" - are you worried about the locks? I've considered shooting everything into a channel and have the receiver handle it, but it seems too high-volume to just use a Feed (which panics if the consumer falls behind), and if locking is an issue then cramming everything up a single channel could be as well (though you would be able to drop information), at which point I don't see the benefit of it.

MBerhault commented:
@tschottdorf I agree a channel wouldn't help much when reaching the limits, it would merely be more idiomatic. That said, the callers don't need anything from the tracer, it's purely a fire-and-forget mechanism. A channel -- or even crazier and probably overkill: goroutines -- would make that clearer.

tschottdorf commented:
I've updated the concept of a `Trace` to be an autonomous object that submits itself on `Finalize()` and am passing them along with the context in most places. No work has been done on figuring out which ones to keep or how to store them yet (or whether `util/tracer` is the right place for this to live in), but this seems like a good first step and I'm generally happy with the informations traces in that form give. PTAL

tschottdorf commented:
The builds for this PR just keep failing with random errors, it's likely since #1482 is fresh in.

petermattis commented on util/tracer/resolver.go:
```diff
@@ -0,0 +1,109 @@
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	buf := bytes.NewBuffer(nil)
```
I usually write this as `var buf bytes.Buffer`.

tschottdorf replied:
done.

petermattis commented on util/tracer/tracer.go:
```diff
@@ -0,0 +1,198 @@
+// util.Feed registered by the Tracer on completion.
+type Tracer struct {
+	origin  string // owner of this Tracer, i.e. Host ID
+	callers *CallResolver
```
Perhaps avoid the allocation of `callers` and make the type `CallResolver`?

tschottdorf replied:
There's only one `Tracer` per Node, so it doesn't matter but I'll give it a try, always happy to get rid of a pointer. CallResolver contains a mutex, so that will probably have to turn into a pointer, which translates into more nil checks, though. We'll see.

petermattis replied:
Well, `CallResolver` could really be private to this package. And then add `CallResolver.init` to perform the initialization. Or maybe it doesn't matter.

petermattis commented on util/tracer/tracer_test.go:
```diff
@@ -0,0 +1,154 @@
+	return id.TraceID()
+}
+
+func TestTracer(t *testing.T) {
```
Would be good to test the `Trace.String()` formatting.

tschottdorf replied:
yes, more tests will be coming when I feel moderately confident that things are somewhat stable (moving to protos and all).

petermattis commented on kv/txn_coord_sender.go:
```diff
@@ -185,6 +187,11 @@ func (tm *txnMetadata) close(txn *proto.Transaction, resolved []proto.Key, sende
 		// effort. We simply fire and forget, each in its own goroutine.
 		if stopper.StartTask() {
 			go func() {
+				// TODO this isn't threaded so the traces may nest randomly.
```
`s/TODO/TODO(tschottdorf)/g`

tschottdorf replied:
done

petermattis commented on storage/range.go:
```diff
@@ -497,18 +501,29 @@ func (r *Range) SetLastVerificationTimestamp(timestamp proto.Timestamp) error {
 func (r *Range) AddCmd(ctx context.Context, call proto.Call, wait bool) error {
 	args, reply := call.Args, call.Reply
 	header := args.Header()
+	// TODO Most requests are ok (those from store.ExecuteCmd) but others get here
```
`s/TODO/TODO(tschottdorf)/g`

tschottdorf replied:
done

petermattis commented on storage/range.go:
```diff
@@ -497,18 +501,29 @@ func (r *Range) SetLastVerificationTimestamp(timestamp proto.Timestamp) error {
+	// directly. Make sure to force enforce ClientCmdID beforehand.
+	trace := tracer.FromCtx(ctx)
+	if trace == nil {
+		trace = r.rm.Tracer().Trace(header)
```
Do you have to `defer trace.Finalize()` here?

tschottdorf replied:
yes, good eye.

petermattis commented on storage/range.go:
```diff
@@ -675,6 +690,16 @@ func (r *Range) addWriteCmd(ctx context.Context, args proto.Request, reply proto
+	if !wait {
+		trace.Event("raft [async]")
+		// Remove the trace from the context. Raft will process asynchronously.
+		// TODO give it a new trace.
```
`s/TODO/TODO(tschottdorf)/g`

tschottdorf replied:
done.

petermattis commented on util/tracer/context.go:
```diff
@@ -0,0 +1,38 @@
+
+// contextKey is the key claimed for storing and retrieving a Trace from
+// a context.Context.
+const ContextKey = "_tracer"
```
`s/ContextKey/contextKey/g`

tschottdorf replied:
done.

petermattis commented on util/tracer/tracer.go:
```diff
@@ -0,0 +1,198 @@
+// marks the end of the Trace, at which point it publishes itself to an
+// associated `util.Feed`. A request may create multiple Traces as it passes
+// through different parts of a distributed systems.
+// A Trace is not safe for concurrent access.
```
How are you guaranteeing that concurrent access does not occur? For example, are traces local to a goroutine?

tschottdorf replied:
yes, simply no passing Traces into a new Goroutine for now. It usually follows an `(args, reply)` pair, which are also not safe for concurrent access, so I haven't had much trouble with it. Of course there are some trickier portions around Raft, but the race detector immediately finds those things since all our tests send requests.
Of course we can always slap a Lock on a Trace and make it concurrent. But the outcome isn't desirable: You get trace entries from different actors all muddled together. That was my initial design and it wasn't pretty to look at.

petermattis commented:
LGTM though I'm a bit concerned about the lack of thread-safety for `tracer.Trace`. Looks very easy to accidentally use it concurrently and then things will blow up. Any thoughts on how to protect against that?

tschottdorf commented:
Getting ready to merge. I added `trace.Fork()` to make it easy to fork off a Trace when spawning a goroutine (helpful if the original Traceable isn't available any more).

Nothing's really done with the Traces yet except that they're merged at verbosity two (best viewed with `--vmodule=monitor=2`). They're already helpful for seeing what requests tests or the bank example do and how long each thing takes.

tschottdorf commented:
Some more example traces are here: https://gist.github.com/tschottdorf/96756ddbaad4aafe7121

mrtracy commented on util/tracer/tracer.go:
```diff
@@ -0,0 +1,212 @@
+
+// A Tracer is used to follow requests across the system (or across systems).
+// Requests must implement the Traceable interface and can be traced by invoking
+// Create(), which returns a Trace object initialized to publish itself to a
```
The method appears to be named NewTrace(), not Create().

tschottdorf replied:
done.

mrtracy commented on server/status/feed.go:
```diff
@@ -91,6 +92,7 @@ func (nef NodeEventFeed) CallComplete(args proto.Request, reply proto.Response)
 // listen for events published by nodes.
 type NodeEventListener interface {
 	OnStartNode(event *StartNodeEvent)
+	OnTrace(event *tracer.Trace)
```
This is fine for now, but can you add a TODO to break this out into its own "TraceEventListener"?  Actually, i'm trying to figure out if this is the best way to handler these listeners; as is, you can't use multiple event listeners on the same subscription.

tschottdorf replied:
done.

mrtracy commented on kv/txn_coord_sender.go:
```diff
@@ -184,16 +187,16 @@ func (tm *txnMetadata) close(txn *proto.Transaction, resolved []proto.Key, sende
 		// effort. We simply fire and forget, each in its own goroutine.
 		if stopper.StartTask() {
-			go func() {
+			go func(ctx context.Context) {
```
Is there any advantage to adding parameters to the immediately invoked function, rather than just using them as a closure?

tschottdorf replied:
you can't accidentally change the value from outside. But here it's for emphasis only.

mrtracy commented on util/tracer/context.go:
```diff
@@ -0,0 +1,38 @@
+
+// contextKey is the key claimed for storing and retrieving a Trace from
+// a context.Context.
+const contextKey = "_tracer"
```
The [examples](http://blog.golang.org/context) given by Google on how to do this use named integer types, like:

```
type key int

const contextKey key = 0
```

No idea if that would increase performance at all, but its also makes it impossible to collide with a key from anywhere else.

tschottdorf replied:
ah, neat. Done.

mrtracy commented:
The naming here is hard to parse sometimes; `Trace` and `Tracer` are very similar, and the package name is also `tracer`.  I'm honestly surprised that variables like `tracer *tracer.Tracer` aren't running afoul of our linters.

tschottdorf commented:
yeah, I tend to agree. Open to suggestions.

tbg committed:
basic request tracing

added the `tracer` package, providing facilities to trace requests through
the cluster. Nothing is stored permanently yet, but running with
  `--vmodule=monitor=2`
logs all request traces. Dealing with the created traces is the next step.

reworked API based on the suggestions from the first proposal: a Trace is
autonomous, and submits itself on `Finalize()`. For tracing requests, have

- `trace.Event("x happened")` and
- `fn := trace.Epoch("stuff begins now"); fn() // stuff ends now`, with its
  most convenient use case `defer trace.Epoch("x")()`.

Epochs and Events generate the same entries, but the former with nonzero
duration. Each entry has a depth, so nested epochs can be displayed nicely:

single-request increment, node side. Note that the total duration of the call
(609.166us) is right at the top; the nested epochs below divvy it up.
```
ID                 Origin              Ts                 Dur          Desc                    File
1435427552[...]    macts.local:8001    13:52:32.200487    609.166µs    node                    server/node.go:564
1435427552[...]    macts.local:8001    13:52:32.200515    576.619µs    ·executing Increment    storage/store.go:1271
1435427552[...]    macts.local:8001    13:52:32.200554    538.349µs    ··range                 storage/range.go:517
1435427552[...]    macts.local:8001    13:52:32.200561    0            ···write path           storage/range.go:526
1435427552[...]    macts.local:8001    13:52:32.200576    515.994µs    ···raft                 storage/range.go:703
1435427552[...]    macts.local:8001    13:52:32.200583    48.923µs     ····proposing           storage/range.go:765
1435427552[...]    macts.local:8001    13:52:32.200971    107.385µs    ····executing           storage/range.go:799
```

`EndTransaction` on the client gateway side:
```
I0627 13:52:32.785017   53305 monitor.go:211] received trace:
ID          Origin              Ts                 Dur            Desc                             File
2f2ecd58    macts.local:8001    13:52:32.743443    11.079554ms    coordinated sender               kv/txn_coord_sender.go:324
2f2ecd58    macts.local:8001    13:52:32.743451    1.57359ms      ·sending EndTransaction          kv/dist_sender.go:638
2f2ecd58    macts.local:8001    13:52:32.743455    5.607µs        ··meta descriptor lookup         kv/dist_sender.go:675
2f2ecd58    macts.local:8001    13:52:32.743483    0              ··sending to macts.local:8001    rpc/send.go:178
```

Transactional `Get` which hits an intent, attempts to resolve it
asynchronously, backs off and gets an error not retryable at the store level.
```
ID          Origin              Ts                 Dur             Desc                   File
3de447e3    macts.local:8001    13:52:32.888115    139.456068ms    node                   server/node.go:564
3de447e3    macts.local:8001    13:52:32.888125    126.494154ms    ·executing Get         storage/store.go:1271
3de447e3    macts.local:8001    13:52:32.888128    110.367783ms    ··range                storage/range.go:517
3de447e3    macts.local:8001    13:52:32.888131    110.364629ms    ···read path           storage/range.go:523
3de447e3    macts.local:8001    13:52:32.998509    16.098773ms     ··intent resolution    storage/store.go:1343
3de447e3    macts.local:8001    13:52:33.014457    150.888µs       ···range               storage/range.go:517
3de447e3    macts.local:8001    13:52:33.014460    0               ····write path         storage/range.go:526
3de447e3    macts.local:8001    13:52:33.014489    0               ····raft [async]       storage/range.go:698
3de447e3    macts.local:8001    13:52:33.014625    165.802µs       ·backoff               storage/store.go:1264
3de447e3    macts.local:8001    13:52:33.014796    12.765902ms     ·executing Get         storage/store.go:1271
3de447e3    macts.local:8001    13:52:33.014802    12.759787ms     ··range                storage/range.go:517
3de447e3    macts.local:8001    13:52:33.014805    12.755809ms     ···read path           storage/range.go:523
3de447e3    macts.local:8001    13:52:33.027566    0               ·not successful        storage/store.go:1266
```

TODO:

- convert to protos
- metadata handling? (ex: peer at request boundary)
- figure out persisting and viewing Traces (as discussed in PR)
